Replace deallocation observer

SamSchott opened the pull request:
This PR closes #185 by replacing the deallocation observer mechanism. Its still work in progress but I am posting here because I'd like some feedback from @dgelessus on a couple of questions.

The mechanism for handling new attributes with Python instances as values is as follows:

1. When the Python instance `pyo` is assigned through `__setattr__`, a new `WrappedPyObject` is created to "wrap" that object. Practically, the `WrappedPyObject` manually increments the refcount of `pyo` and stores the address to `pyo` in an ivar.
2. This `WrappedPyObject` is associated with the original ObjCInstance through `objc_setAssociatedObject` which holds the only reference to it. The lifecycles of the ObjCInstance and the `WrappedPyObject` are therefore tied together.
3. When the ObjCInstance is deallocated, so is the `WrappedPyObject`. This will cause the refcount of `pyo` to be decremented in the overridden `WrappedPyObject.dealloc` method.

I've also updated the tests to ensure that all Python objects are indeed destroyed when they should be. This has been working quite well in my manual testing but I am still uncertain about a couple of implementation details. The questions are inline.

A larger question is regarding performance. Creating a new `WrappedPyObject` for each instance attribute can potentially be costly. But I don't really see an easy way around that. What do you think?

Edit: @freakboy3742 is of course also very welcome to comment!

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

SamSchott commented on rubicon/objc/api.py:
```diff
@@ -851,17 +828,31 @@ def __setattr__(self, name, value):
+
+                # Wrap the Python object as an WrappedPyObject instance.
+                # A reference will be retained as long as the WrappedPyObject is alive.
+                wrapper = WrappedPyObject.alloc().initWithObjectId_(id(value))
```
This is more legible for the time being but should probably be replaced by the appropriate `send_message` calls.

SamSchott replied:
I've replaced this with the appropriate `send_message` calls.

SamSchott commented:
There is one more issue which I have come across with toga: all example apps segfault when executing a callback, for instance on a button press. The stacktrace is not very helpful and I am not sure if this is an issue with the changes here or with toga itself. Toga may have implicitly relied on (most) ObjCInstances to live forever.

SamSchott commented on rubicon/objc/api.py:
```diff
@@ -803,36 +795,21 @@ def __getattr__(self, name):
-            # so that the only remaining reference is released
-            # when the associated object is deallocated.
-            send_message(observer, 'release', restype=None, argtypes=[])
+        send_message(pyo_wrapper, 'release', restype=None, argtypes=[])
```
Never mind, the explicit call to release is not necessary after all. It appears that the object returned by `libobjc.objc_getAssociatedObject()` is autoreleased. Since the tests don't use an `NSAutoreleasePool`, those references where never cleaned up.

SamSchott commented:
Turns out that the explicit `release` call in `__getattr__()` was not necessary after all and was causing the crashes in toga.

SamSchott commented:
Alright, this should be ready review :)

dgelessus commented on rubicon/objc/api.py:
```diff
@@ -1554,42 +1552,47 @@ def __subclasscheck__(self, subclass):
-# ObjCInstance._keep_alive_objects dictionary.
+# When a Python object is assigned to a new ObjCInstance attribute, the Python
+# object should be kept alive for the lifetime of the ObjCInstance. This is
+# done by wrapping the Python object as a WrappedPyObject that manages increments
+# the reference count during assignment and decrements it the the WrappedPyObject
```
```suggestion
# done by wrapping the Python object as a WrappedPyObject that increments
# the reference count during assignment and decrements it when the WrappedPyObject
```

dgelessus commented on rubicon/objc/api.py:
```diff
@@ -1554,42 +1552,47 @@ def __subclasscheck__(self, subclass):
+    class WrappedPyObject(NSObject):
 
-        observed_object = objc_ivar(objc_id)
+        wrapped_pointer = objc_ivar(c_long)
```
Would `c_void_p` work here instead, to make it clear that it's a pointer?

SamSchott replied:
Yes.

dgelessus commented on rubicon/objc/api.py:
```diff
@@ -1554,42 +1552,47 @@ def __subclasscheck__(self, subclass):
             self = send_message(self, 'init', restype=objc_id, argtypes=[])
             if self is not None:
-                set_ivar(self, 'observed_object', anObject)
+                pyo = cast(address, py_object)
+                pyo_ptr = cast(id(pyo.value), POINTER(PyObject))
```
`id(pyo.value)` should be the same as `address`, right? Would it be possible to remove the `pyo` variable and use `address` directly here instead? Or is the cast to `py_object` important for reference counting reasons (or something else)?

```suggestion
                pyo_ptr = cast(address, POINTER(PyObject))
```

SamSchott replied:
Good catch, `id(pyo.value)` and `address` are indeed the same. This roundabout way is just a leftover from previous code which I didn't clean up properly.

dgelessus commented on rubicon/objc/api.py:
```diff
@@ -1554,42 +1552,47 @@ def __subclasscheck__(self, subclass):
-            ObjCInstance._keep_alive_objects.pop(anObject.value, None)
+            address = get_ivar(self, 'wrapped_pointer')
+            if address.value:
+                pyo = cast(address.value, py_object)
+                pyo_ptr = cast(id(pyo.value), POINTER(PyObject))
```
Like above - is the `pyo` variable needed here?

```suggestion
                pyo_ptr = cast(address.value, POINTER(PyObject))
```

dgelessus commented on rubicon/objc/api.py:
```diff
@@ -1554,42 +1552,47 @@ def __subclasscheck__(self, subclass):
+            if address.value:
+                pyo = cast(address.value, py_object)
+                pyo_ptr = cast(id(pyo.value), POINTER(PyObject))
+                pythonapi.Py_DecRef(pyo_ptr)
```
I think `Py_DecRef`'s `restype`/`argtypes` aren't set anywhere in the current version of the PR. `ctypes_patch` declares `Py_IncRef`, but not `Py_DecRef` - and as explained above, I would like to avoid depending on `ctypes_patch` for these definitions.

SamSchott replied:
Sure, I can add those definitions in a new module for consistency.

SamSchott replied:
Done.

dgelessus commented on rubicon/objc/api.py:
```diff
@@ -1598,8 +1601,11 @@ def finalize(self, cmd) -> None:
-            ObjCInstance._keep_alive_objects.pop(anObject.value, None)
+            address = get_ivar(self, 'wrapped_pointer')
+            if address.value:
+                pyo = cast(address.value, py_object)
+                pyo_ptr = cast(id(pyo.value), POINTER(PyObject))
```
```suggestion
                pyo_ptr = cast(address.value, POINTER(PyObject))
```

dgelessus commented on tests/test_core.py:
```diff
@@ -1163,21 +1167,36 @@ def test_objcinstance_python_attribute(self):
+        thing = NSObject.alloc().init()
+
+        # create NSAutoreleasePool to clean up any leftover Obj-C instances
+        pool = NSAutoreleasePool.alloc().init()
```
This isn't something that needs to be done in this PR, but ideally Rubicon should provide a proper API for autorelease pools, probably as a context manager (`with objc.autoreleasepool(): ...`).

In normal Objective-C code, `NSAutoreleasePool` is more or less superseded by `@autoreleasepool` blocks, and when using ARC, `@autoreleasepool` is the only option - see the warning in [the `NSAutoreleasePool` docs](https://developer.apple.com/documentation/foundation/nsautoreleasepool?language=objc). Internally, `@autoreleasepool` blocks are translated to calls to the Objective-C runtime functions `objc_autoreleasePoolPush` and `objc_autoreleasePoolPop`, which are undocumented and not declared in Apple's headers, but [seem to be widely known](https://stackoverflow.com/q/15014152). It's *probably* fine if we keep using `NSAutoreleasePool` for now - it's not officially deprecated, and we aren't affected by ARC restrictions. But for future-proofing it would be better to define our own high-level API for autorelease pools that uses the appropriate mechanism under the hood. (Again, not something that needs to be done in this PR.)

SamSchott replied:
Yes, an `autoreleasepool` context manager would be a good idea. A lot of code that uses rubicon-objc will likely also run a Cocoa event loop which will create and drain pools for each cycle but there of course will remain use cases for manually managing an autorelease pool.

I'm not so sure about using undocumented Objective-C runtime functions. Could this potentially lead to issues with app review if they are considered private? I have never gone through the review process myself...

dgelessus replied:
I've copied the discussion into a separate issue and replied to your question there: #210

dgelessus commented on tests/test_core.py:
```diff
@@ -1163,21 +1167,36 @@ def test_objcinstance_python_attribute(self):
+        # Test deleting Python attribute.
+
+        del thing.python_object_1
+
+        with self.assertRaises(AttributeError):
+            thing.python_object_1
+
     def test_objcinstance_python_attribute_keep_alive(self):
         """Python attributes on an ObjCInstance are kept even if the object temporarily has no Python references."""
 
-        Example = ObjCClass("Example")
-        example = Example.alloc().init()
-        Thing = ObjCClass("Thing")
-        thing = Thing.alloc().init()
+        example = NSObject.alloc().init()
+        thing = NSObject.alloc().init()
```
I think this changes the behavior of the test quite significantly, along with the change below from `example.setThing(thing)` to `example.thing = thing`. Previously, it tested that:

* If the Objective-C object `thing` has Python attributes
* and it is stored inside an Objective-C property on another object (`example`)
* and `thing` temporarily has *no* references from Python
* once `thing` is retrieved from the Objective-C property, it still has its Python attributes

The new test code is similar, but it stores `thing` in a *Python property* on an Objective-C object, and not in a real Objective-C property. The difference is relevant because when `thing` is stored in a Python property, its Python `ObjCInstance` is kept alive. This test is supposed to test the case where the `ObjCInstance` is temporarily destroyed, but the underlying Objective-C object continues to exist, and is then later used from Python again.

The test as it's written now is probably useful to have as well, but it should be a separate test in addition to the old one, not a replacement.

dgelessus reviewed (commented):
Thank you for the PR, and sorry for the late review!

Interesting approach, I hadn't thought about storing the Python objects inside Objective-C associated objects. My thought would have been adding a variant of `objc_property` that stores a Python object, and to require users to declare all Python object properties that they want to use. But this implementation using associated objects is much nicer - it no longer requires keeping `ObjCInstance`s alive, but still allows setting arbitrary Python attributes on any Objective-C object. That should mean that no existing user code needs to be changed, which is always a good thing.

SamSchott commented on rubicon/objc/api.py:
```diff
@@ -16,6 +16,7 @@
     Class, SEL, add_ivar, add_method, ensure_bytes, get_class, get_ivar, libc, libobjc, objc_block, objc_id,
     objc_property_attribute_t, object_isClass, set_ivar, send_message, send_super,
 )
+from .ctypes_patch import PyObject
```
That's a fair point.

SamSchott commented on rubicon/objc/api.py:
```diff
@@ -851,17 +827,39 @@ def __setattr__(self, name, value):
+
+                # Set the Python value as an associated object. This will release
+                # any previous wrapper object with the same key.
+                libobjc.objc_setAssociatedObject(self, SEL(name), wrapper, 0x301)
```
Agreed, it's better to have a namespace for those automatically defined selectors.

SamSchott commented on rubicon/objc/api.py:
```diff
@@ -851,17 +827,39 @@ def __setattr__(self, name, value):
-                # every ObjCInstance with additional Python attributes is "kept alive" on the Python side.
-                # This prevents the ObjCInstance from being destroyed while the Objective-C object still exists.
-                self._ensure_keep_alive()
+            elif name.startswith("_"):
```
I did run into trouble with the `_needs_release` attribute previously but cannot reproduce it now. As you say, it will already exist in the `__dict__` and is therefore caught by the previous check. Pending any upcoming issues, I'll remove the special handling of private attributes.

SamSchott commented:
Thanks for the diligent review! I've addressed most points but still need to move all CPython API definitions to a separate module.

In the meantime, I came across two questions related to freeing unreferenced objects:

1. I tried to adapt the original `test_objcinstance_python_attribute_keep_alive` test to release all Python objects at the end by deleting `thing` and `example` and using an autorelease pool. Somehow, I could not get it to work. Is there anything about the `Example` and `Thing` classes which would prevent the Python objects or the classes themselves from being released?

2. While this PR *should* fix memory leaks in toga, it currently does not. The reason for this are circular references introduced in the Objective-C side. This is a common pattern throughout toga-cocoa:

    ```Python
    from rubicon.objc import NSObject


    class InterfaceObject()
        pass


    class ImplObject()
        pass


    interface = InterfaceObject()
    interface._impl = ImplObject()
    interface._impl.native = NSObject.alloc().init()

    interface._impl.native.interface = interface
    interface._impl.native.impl = interface._impl

   del interface  # should release the entire spagetti but does not
    ```

    The resulting circular references are obviously not garbage collected because Python does not know who is holding on to the remaining references of `interface` and `impl`. Writing our own garbage collection is a really bad idea. Ideally, one would like to declare weak properties on the `NSObject` to store arbitrary Python objects. Do you think this is feasible?

    Introducing such circular references somehow strikes me as a bad design pattern in the first place and maybe it can be avoided. AFAIK, the `interface` and `impl` attributes are only used when `NSObject` is its own target or delegate and needs access to the interface or implementation layer in callbacks. This could be solved by writing separate delegates instead of subclassing the original objects. Since `setDelegate()` does not take ownership of the object, we avoid the reference cycle. @freakboy3742, what do you think?

dgelessus commented:
re. the CPython APIs: sorry for going back and forth on this, but it would actually be better to only pull out those parts of `ctypes_patch` that are part of the documented CPython API and thus safe to use (namely `PyTypeObject`, `PyObject`, `PyVarObject`, `Py_IncRef`, and `Py_DecRef`). All other declarations (`PyDict_Type`, `PyDictObject`, `mappingproxyobject`, `ffi_type`, `GETFUNC`, `SETFUNC`, and `StgDictObject`) are undocumented and possibly unstable, so it would be better to leave those in `ctypes_patch`, together with the other possibly dangerous code.

Also, I think it would be better to name the submodule something other than `pythonapi`, because `ctypes` already uses the name `pythonapi` for its predefined library object for the CPython API. The two names aren't actually conflicting, because the code uses the right combination of `from`-`import`s and only imports one of the two `pythonapi`s directly. But IMO it's still a bit confusing and possibly error-prone to have the module and the library named exactly the same. Maybe the module could be named `pythonapi_defs` or something like that? Then it would be clear that the two are related, but not the same.

> I tried to adapt the original `test_objcinstance_python_attribute_keep_alive` test to release all Python objects at the end by deleting `thing` and `example` and using an autorelease pool. Somehow, I could not get it to work. Is there anything about the `Example` and `Thing` classes which would prevent the Python objects or the classes themselves from being released?

Hmm, not sure what could be going on there... There's always the problem that *officially* Python doesn't guarantee when a Python object is destroyed once it's no longer referenced - but in practice the way that you're doing it (`del` all references to the object, then `gc.collect()`) should be reliable enough. It's especially strange that (if I understand you correctly) the object gets deleted if you use `NSObject` but not if you use `Example` and `Thing`.

> The resulting circular references are obviously not garbage collected because Python does not know who is holding on to the remaining references of `interface` and `impl` on the Objective-C side. Writing our own garbage collection is a really bad idea.

Agreed that we shouldn't try to build our own garbage collection. But sadly I think there's also no way we could make Python's garbage collection properly understand references on the Objective-C side, especially when it comes to more complex reference cycles like "Python object -> ObjC object 1 -> ObjC object 2 -> Python object".

> Ideally, one would like to declare weak properties on the `NSObject` to store arbitrary Python objects. Do you think this is feasible?

That's probably the best solution. In the Objective-C -> Python direction it should be easy to do by just storing a Python `weakref` as an attribute on the Objective-C object. The other direction could be implemented using something similar to the existing `DeallocationObserver` - although Objective-C also has native support for weak references, which could probably be used to implement Python -> Objective-C weakrefs more efficiently and without another custom associated object.

dgelessus commented:
Thinking about the CPython API thing some more - we could completely avoid calling the CPython API if we keep using the existing `_keep_alive_objects` solution for keeping objects alive. Basically, `pythonapi.Py_IncRef(pyo_ptr)` could be replaced with `ObjCInstance._keep_alive_objects[id(pyo)] = pyo`, and `pythonapi.Py_DecRef(pyo_ptr)` with `del ObjCInstance._keep_alive_objects[id(pyo)]`. The difference to before would be that now we only keep the individual attribute values alive, and not the entire `ObjCInstance` object as before.

SamSchott commented:
No worries about the Python API, I fully understand your points about the module name and about not exposing undocumented declarations.

> Thinking about the CPython API thing some more - we could completely avoid calling the CPython API if we keep using the existing _keep_alive_objects solution for keeping objects alive.

Possibly yes, but we would also have to store the reference count for the Objective-C side and only remove the Python object from the dict once it drops to zero. This would handle cases where a Python object is referred to from multiple Objective-C instances. Basically our own simple reference counting. The advantage of this approach is that conventional methods of finding memory leaks such as `gc. get_referrers()` would at least show that the references were created by rubicon-objc. I'd be happy with either solution.

> It's especially strange that (if I understand you correctly) the object gets deleted if you use NSObject but not if you use Example and Thing.

Yes, this is what confused me as well. I also didn't see any issues with other Cocoa classes such as `NSButton`, etc.

Regarding storing weak references to Python objects, I am not sure I fully understand your suggestions, especially regarding the  Python -> Objective-C direction. Do you mean creating weak references to Objective-C objects?

Broadly, I can see the following APIs:

1. Allow `objc_property` to accept Python objects and provide an additional argument specifying weak vs strong references. Strong references to Python objects can again be implemented by managing the ref count whereas weak references would just leave the ref count alone.
2. Provide some kind of `weakref` wrapper to wrap Python objects when assigning them to Objective-C attributes. This would enable creating attributes with weak references for individual instances instead of declaring them for the class. For example:

    ```Python
    obj = NSObject.alloc().init()
    pyo = object()
    obj.attribute = objc_weakref(pyo)
    assert objc.attribute is pyo
   ```
   
   This `objc_weakref` would just tell the `WrappedPyObject` not to manage the ref counts. This approach might however be confusing.
3. Don't deal with the issue at all in rubicon-objc and require users to manually use Python's `weakref` module.

I tend to prefer the first option, also because it allows storing weak references to both Python and Objective-C objects. What do you think?

dgelessus commented:
> Possibly yes, but we would also have to store the reference count for the Objective-C side and only remove the Python object from the dict once it drops to zero. This would handle cases where a Python object is referred to from multiple Objective-C instances.

Ah, good point, I forgot about that case. I think this could also be solved by making the key a tuple that contains both the address of the referenced Python object and the address of the referencing `WrappedPyObject`. That would avoid having to write our own reference counting.

IMO a Python-based solution like this would be better than calling `Py_IncRef`/`Py_DecRef` via `ctypes`. It saves us the trouble of having to interact with the C API, it's easier to ensure that the references are added/deleted correctly, and as you've already mentioned it makes it easier to trace the references while debugging.

> Regarding storing weak references to Python objects, I am not sure I fully understand your suggestions, especially regarding the Python -> Objective-C direction. Do you mean creating weak references to Objective-C objects?

When I replied, I wasn't sure which direction you were asking about, so I wrote a short reply about each direction 😄 But re-reading your comment again now I think you were only talking about adding weak references from an Objective-C object to a Python object, not the other way around, so I guess you can ignore the second half of my comment 🙂

I think the simplest solution for now is to have no special support for weak references in Rubicon. In cases where a reference to a Python object needs to be weak, you can manually wrap the object in a `weakref` before storing it as an attribute. The overhead of doing it manually shouldn't be very big - you just need to replace `obj.attr = value` with `obj.attr = weakref.weakref(value)` when setting the attribute, and replace `obj.attr` with `obj.attr()` when using it.

I agree that it would be nice if we could declare Objective-C properties for storing Python objects (i. e. `objc_property(py_object)`). Then we could add native support for weak references in `objc_property` as you're suggesting. But that would require some additional work on `objc_property` that isn't really related to this PR (replacing `DeallocationObserver`), so IMO it would be better to do that as a separate PR in the future.

SamSchott commented:
> I think this could also be solved by making the key a tuple that contains both the address of the referenced Python object and the address of the referencing WrappedPyObject

Agreed, that solution is easier to manage and debug than using the C API. It might also come in handy when writing tests, though we probably mostly want to test the API (i.e., that Python objects are retained and freed correctly) and not the implementation.

Regarding weak references, yes, I was only talking about weak references in the Objective-C -> Python direction, though the reverse could certainly be useful as well. But in any case, I agree that this is best done in other PR. I just raised it here because I was a bit disappointed when realising that this PR does not solve the memory leaks in toga-cocoa :(

dgelessus reviewed (approved):
Alright, I think that's everything addressed now! Thanks for the PR - even if it doesn't directly fix the memory management issues you're running into, it makes the Python attribute storage more robust, and will probably help with implementing `objc_property(py_object)` in the future.

Before merging - I'm thinking if it would make sense to drop the commits be7e48c5a4bf3b63f3008f533a908e235f157017 and dbf540cf8868ea375e3ebbb839717d8fd4d0b526 from the PR. The second commit just reverts the first one, and they are quite close to each other in the history, so it should be possible to safely remove them. This would help with keeping the Git history (and especially `git blame`) a bit cleaner in the long term.

I don't know how familiar/comfortable you are with interactive rebase and rewriting Git history - you can make the changes yourself if you like, but I can also do while merging the PR if you prefer.

SamSchott commented:
Thanks for the thorough reviews and for bearing with my while I was finding my way through ctypes and the ObjC runtime!

Yeah, there are a lot of commits in the history just from trying different approaches and reverting them. I've squashed them all into a single commit for the changes to the packages itself and left separate commits for the tests and towncrier entry.
